Add/Remove image secret entries

jhadvig opened the pull request:
Implementing logic for adding/removing image secret credentials, since the secret can contain more then one credentials. Demo:
![secrets okd](https://user-images.githubusercontent.com/1668218/44657345-ce223080-a9fc-11e8-950e-2721f07cd95a.gif)

Used design suggested by @beanh66 (please help review) in https://github.com/openshift/console/pull/205#issuecomment-406587544 with small changes:
- The *Remove Credentials* button also contains `fa-minus-circle` icon, similarly to the *Add Credentials* button.
- All the entries have *Remove Credentials* button above then, aligned to the right side
- If there is just one entry, the *Remove Credentials* button will be hidden, since the form is meant for editing the secret, not for deletion.

Had to add additional `uid` field to the `ConfigEntryFormState` so React will re-render only updated entries. Could not use *index* of the entry in the array since thats not unique(since the entries on different positions can be now removed) or any other attribute of the entry(address, name, pw, mail), since that not unique as well and can be changed.

Tested the changes for both image secret formats and should work as described without issues.

One thing that I'm hesitating to change are the names that contain `entry` string. Since the Add/Remove buttons are referring to `credentials`. Kinda like the `entry` string in the code more. Also the code would require much more changes, and if we decide to go with the change I would rather do it in a separate commit, so the changes are obvious.

@spadgett PTAL

jhadvig commented:
Also replaced selectbox of  the *Authentication Type* with `Dropdown` component (not visible on the demo gif)
![1](https://user-images.githubusercontent.com/1668218/44669531-6847a000-aa20-11e8-9083-1c24d53482a5.png)

robszumski commented:
> Used design suggested by @beanh66 (please help review) in #205 (comment) with small changes:

It looks like you missed the border between them, which is exactly what I was thinking this needed when I first saw it. Can we add that back? cc @beanh66

jhadvig commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -419,22 +428,53 @@ class CreateConfigSubform extends React.Component<CreateConfigSubformProps, Crea
   render() {
-    const secretEntriesList = _.map(this.state.secretEntriesArray, (entry, index) => <ConfigEntryForm key={index} id={index} entry={entry} onChange={this.onDataChanged} />);
+    const secretEntriesList = _.map(this.state.secretEntriesArray, (entry, index) => {
+      return <div className="co-create-image-secret__form-wrapper" key={`${entry.uid}`}>
```
Read that doc, but thought that since there are two siblings, would be better to hook them up with proper key names - `${entry.uid}-form` & `${entry.uid}-link`, since they will will have to be different anyways
But they can be names as suggested :)

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -419,22 +428,53 @@ class CreateConfigSubform extends React.Component<CreateConfigSubformProps, Crea
-    const secretEntriesList = _.map(this.state.secretEntriesArray, (entry, index) => <ConfigEntryForm key={index} id={index} entry={entry} onChange={this.onDataChanged} />);
+    const secretEntriesList = _.map(this.state.secretEntriesArray, (entry, index) => {
+      return <div className="co-create-image-secret__form-wrapper" key={`${entry.uid}`}>
+        {_.size(this.state.secretEntriesArray) > 1 && <button className="btn btn-link remove-entry-link" key={`${entry.uid}-btn`} onClick={(e) => this.removeEntry(index, e)}>
```
Actually I don't think `key` is needed here

jhadvig commented:
@spadgett I've addressed your comments in two commits, so they are more readable (will squash before merge)
The last added commit **drowpdown width** addresses the generic fix  for [your comment](https://github.com/openshift/console/pull/459#discussion_r213141757). Tested the change on the role-binding/name-value-editor/secrets and works as before.

rhamilto commented on frontend/public/components/_dropdown.scss:
```diff
@@ -255,7 +256,7 @@ $color-bookmarker: #DDD;
 
 .btn--dropdown {
-  max-width: 100%;
+  width: 100%;
```
Agreed.  This class is arguably misusing BEM.  Should probably be `.btn-dropdown` then you could add a `.btn-dropdown--full-width` or something along those lines.

jhadvig replied:
@rhamilto makes sense ! :+1:

jhadvig commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -216,7 +220,7 @@ class ImageSecretForm extends React.Component<ImageSecretFormProps, ImageSecretF
       dataKey: getImageSecretKey(this.props.secretType),
       stringData: parsedData,
-      uploadConfig: 'false',
+      uploadConfig: 'credentials',
```
`authType` :+1:

jhadvig commented:
@spadgett I've updated the Pr based on tyour and @rhamilto comments.
I've replaced the `.btn--dropdown` with `.btn-dropdown` so it aligns with BEM, but instead of adding  `.btn-dropdown--full-width` I've added `.dropdown--full-width` so it's  button and menu children can use the `width:100%`

PTAL

spadgett commented on frontend/public/components/secrets/create-secret.tsx:
```diff
@@ -70,6 +70,10 @@ const generateSecret = () => {
 
+const generateUID = (prefix) => {
+  return _.uniqueId(prefix);
+};
```
I think my comment got lost earlier, but I don't see a reason to wrap `_.uniqueId`. We're just calling it with the same parameter.

spadgett commented:
@rhamilto do you mind reviewing the styles?

spadgett commented:
jenkins rebuild

jhadvig commented:
jenkins rebuild

jhadvig commented:
jenkins rebuild
